### PR TITLE
Enable support for unboxing [K: [Unboxable]]

### PR DESCRIPTION
**Pull request by johannth**
This PR enables support for unboxing values of type `[K: [Unboxable]]`

Example:

```swift
struct Answer: Unboxable {
    let id: String 

    init(unboxer: Unboxer) {
        self.id = unboxer.unbox("id")
    }
}

struct Game: Unboxable {
    let answersByPlayer: [String: [Answer]]

    init(unboxer: Unboxer) {
        self.answersByPlayer = unboxer.unbox("answers")
    }
}
````

We can support both dictionaries of simple values and dictionaries of `Unboxable` by splitting  `unbox<K: UnboxableKey, V: CollectionType>() -> [K: V]` into

```swift
public func unbox<K: UnboxableKey, V: DictionaryLiteralConvertible>(key: String, isKeyPath: Bool = false) -> [K : V]
public func unbox<K: UnboxableKey, V: DictionaryLiteralConvertible>(key: String, isKeyPath: Bool = false) -> [K : V]?
public func unbox<K: UnboxableKey, V: UnboxableRawType>(key: String, isKeyPath: Bool = false, allowInvalidElements: Bool = false) -> [K : [V]]
public func unbox<K: UnboxableKey, V: UnboxableRawType>(key: String, isKeyPath: Bool = false, allowInvalidElements: Bool = false) -> [K : [V]]?
public func unbox<K: UnboxableKey, V: Unboxable>(key: String, isKeyPath: Bool = false, allowInvalidElements: Bool = false) -> [K : [V]]
public func unbox<K: UnboxableKey, V: Unboxable>(key: String, isKeyPath: Bool = false, allowInvalidElements: Bool = false) -> [K : [V]]?
```

I'm new to this codebase (and Swift generics) so please point me towards an easier implementation if you see it.

The original implementation didn't work because we had to cast from `AnyObject` to `T = [K: [Unboxable]]` in `resolveOptionalValueForKey` which didn't work.

### CODE REVIEW

**Comment by JohnSundell** [comment]
Hi @johannth, thanks for your PR! Will take a more thorough look at it tomorrow, initially looks good - but want to think if there's a more generic way to do this than creating all these new overloads. Can't think of one now though :)

File: Sources/Unbox.swift [hunk 1]
```diff
@@ -526,16 +526,44 @@ public class Unboxer {
         return UnboxValueResolver<[String : V]>(self).resolveDictionaryValuesForKey(key, isKeyPath: isKeyPath, required: false, allowInvalidElements: false, valueTransform: { $0 })
     }
     
-    /// Unbox a required Dictionary containing collections
-    public func unbox<K: UnboxableKey, V: CollectionType>(key: String, isKeyPath: Bool = false) -> [K : V] {
+    /// Unbox a required Dictionary containing dictionaries
+    public func unbox<K: UnboxableKey, V: DictionaryLiteralConvertible>(key: String, isKeyPath: Bool = false) -> [K : V] {
```
Comment (JohnSundell) [inline Sources/Unbox.swift]: Make this constraint match the one on `UnboxValueResolver`; so `T: CollectionType, T: DictionaryLiteralConvertible, T.Generator == DictionaryGenerator<T.Key, T.Value>`. That way custom `DictionaryLiteralConvertible` objects are less likely to be matched.

File: Sources/Unbox.swift [hunk 2]
```diff
@@ -526,16 +526,44 @@ public class Unboxer {
...
+    public func unbox<K: UnboxableKey, V: Unboxable>(key: String, isKeyPath: Bool = false, allowInvalidElements: Bool = false) -> [K : [V]] {
+        return UnboxValueResolver<[String: [UnboxableDictionary]]>(self).resolveDictionaryValuesForKey(key, isKeyPath: isKeyPath, required: true, allowInvalidElements: allowInvalidElements) {
+            return try? Unbox($0, context: self.context, allowInvalidElements: allowInvalidElements)
+            } ?? [K: [V]]()
+    }
+
+    /// Unbox an optional Dictionary containing array of Unboxables
+    public func unbox<K: UnboxableKey, V: Unboxable>(key: String, isKeyPath: Bool = false, allowInvalidElements: Bool = false) -> [K : [V]]? {
+        return UnboxValueResolver<[String: [UnboxableDictionary]]>(self).resolveDictionaryValuesForKey(key, isKeyPath: isKeyPath, required: false, allowInvalidElements: allowInvalidElements) {
+                        return try? Unbox($0, context: self.context, allowInvalidElements: allowInvalidElements)
```
Comment (JohnSundell) [inline Sources/Unbox.swift]: This indentation seems a bit off :)

File: Tests/UnboxTests.swift [hunk 3]
```diff
@@ -359,6 +359,138 @@ class UnboxTests: XCTestCase {
...
+                XCTFail("Key was missing from unboxed dictionary")
+            }
+            
+            let unboxedWithoutOptionals: Model = try Unbox([
+                "requiredModelDictionary" : [
+                    "key" : [
+                        ["int" : 31]
+                    ]
+                ]
+                ])
```
Comment (JohnSundell) [inline Tests/UnboxTests.swift]: Incorrect indentation here (and in a few other places with the same formatting).

**Comment by JohnSundell** [comment]
Hey @johannth! Sorry for the late reply. PR looks really good - I'd make the generic constraint on the new `unbox` method a bit more specific, and also there are some formatting issues - but other than that - I'd be happy to merge this in :) You'll need to merge `master` in though and fix some conficts :)

**Comment by johannth** [comment]
@JohnSundell I've rebased with master and made fixes based on your comments. Can you have another look?

**Comment by JohnSundell** [comment]
Looks great, thanks for this @johannth!

**Pull request merged**